compiler/resolution/tuples.cpp

```diff
@@ -0,0 +1,848 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+
+#include "resolution.h"
+
+#include "astutil.h"
+#include "caches.h"
+#include "chpl.h"
+#include "expr.h"
+#include "stmt.h"
+#include "stringutil.h"
+#include "symbol.h"
+
+#include "resolveIntents.h"
+
+#include "view.h"
+
+#include <cstdlib>
+#include <inttypes.h>
+#include <vector>
+#include <map>
+
+struct TupleInfo {
+  TypeSymbol* typeSymbol;
+  // type constructor is in type->defaultTypeConstructor
+  // constructor is in type->defaultInitializer
+  FnSymbol *buildTupleType;
+  FnSymbol *buildStarTupleType;
+  FnSymbol *buildTupleTypeNoRef;
+};
+
+
+static std::map< std::vector<TypeSymbol*>, TupleInfo > tupleMap;
+
+static
+TupleInfo getTupleInfo(std::vector<TypeSymbol*>& args,
+                       BlockStmt* instantiationPoint,
+                       bool noref)
+{
+  TupleInfo& info = tupleMap[args];
+  if (!info.typeSymbol) {
+    SET_LINENO(dtTuple);
+
+    int size = args.size();
+
+    if (size == 0)
+      USR_FATAL(instantiationPoint, "tuple must have positive size");
+
+    ModuleSymbol* tupleModule =
+      toModuleSymbol(dtTuple->symbol->defPoint->parentSymbol);
```

Comment:
eh... it would be constant between calls to the function but I don't actually think putting it in a static variable saves real performance - and at the same time I think it makes the code more brittle. What if for some reason dtTuple changed between calls? (Not that I have any plan for why it would do so)